### PR TITLE
feat(AccordionItem): Allow controlled component

### PR DESCRIPTION
**Pull request by chardos**
Update to  the `AccordionItem` to allow external control via the `isOpen` prop. If the `isOpen` prop is passed, internal state will no longer be used to determine open/closed state.

### CODE REVIEW

File: react/AccordionItem/AccordionItem.test.js [hunk 1]
```diff
@@ -62,15 +66,19 @@ describe('AccordionItem', () => {
...
       );
 
       const renderedElement = ReactTestUtils.renderIntoDocument(element);
       const button = renderedElement.querySelector('button');
 
       ReactTestUtils.Simulate.click(button);
-      expect(mockOnClose).toHaveBeenCalledTimes(1);
+
+      // wait for useEffect
+      requestAnimationFrame(() => {
```
Comment (grahamgoh) [inline react/AccordionItem/AccordionItem.test.js]: hmm isnt this asynchronous , dont u have to await this?

File: react/AccordionItem/AccordionItem.test.js [hunk 2]
```diff
@@ -53,7 +53,11 @@ describe('AccordionItem', () => {
       const button = renderedElement.querySelector('button');
 
       ReactTestUtils.Simulate.click(button);
-      expect(mockOnOpen).toHaveBeenCalledTimes(1);
+
+      // wait for useEffect
+      requestAnimationFrame(() => {
+        expect(mockOnOpen).toHaveBeenCalledTimes(1);
```
Comment (grahamgoh) [inline react/AccordionItem/AccordionItem.test.js]: hmm isnt this asynchronous , dont u have to await this?
same as below

File: react/AccordionItem/AccordionItem.test.js (hunk 1 above)
Comment (grahamgoh) [reply react/AccordionItem/AccordionItem.test.js]: so the test might be finishing before u even run this expect>

File: react/AccordionItem/AccordionItem.test.js (hunk 2 above)
Comment (chardos) [reply react/AccordionItem/AccordionItem.test.js]: Good catch man. Added a `done()` call

Comment (grahamgoh) [reply react/AccordionItem/AccordionItem.test.js]: nice!

File: react/AccordionItem/AccordionItem.js [hunk 3]
```diff
@@ -10,59 +10,65 @@ function AccordionItem({
...
+        setTimeoutHandle,
+        setCurrentHeight,
+        setCssOpacity,
+        setCssOverflow,
+        setCssVisibility
+      });
+    }
+
+    isMounting.current = false;
+  }, [finalIsOpen]);
```
Comment (grahamgoh) [inline react/AccordionItem/AccordionItem.js]: The usage of isMounting to control whether useEffect should be run feels like a code smell to me. Is there a way to do it without using the isMounting check?

Comment (grahamgoh) [reply react/AccordionItem/AccordionItem.js]: should timeoutHandle be part of the array `[finalIsOpen]` as it is only a state use inside the effect lambda?

Comment (chardos) [reply react/AccordionItem/AccordionItem.js]: > The usage of isMounting to control whether useEffect should be run feels like a code smell to me. Is there a way to do it without using the isMounting check?

Currently, I don't see a tidier way of doing this. In the future we could encapsulate it in a custom hook, maybe something like :

```
useEffectOnUpdate(() => {
  // ...
});
```

Comment (chardos) [reply react/AccordionItem/AccordionItem.js]: > should timeoutHandle be part of the array `[finalIsOpen]` as it is only a state use inside the effect lambda?

We only want to the lambda to run when the state of the accordion is changed, e.g. `finalIsOpen`. The timeout handler will indeed change with the accordion state, but adding it to the array is going to trigger a recursive loop.

**Pull request merged**